File: cpp/software/Application.cpp
#pragma once
#include <bits/stdc++.h>

#include "../component/Software.cpp"
#include "../interfaces/Executable.cpp"
#include "../interfaces/Upgradable.cpp"

class Application : public Software, public Executable, public Upgradable
{
protected:
    std::string appName;
    std::string developer;
    std::string supportedOS;

public:
    // Constructors
    Application() : Software(), appName(""), developer(""), supportedOS("") {}
    Application(std::string appName, std::string developer, std::string supportedOS, std::string softwareVersion, std::string licenseKey, double sizeMB)
        : Software(softwareVersion, licenseKey, sizeMB), appName(appName), developer(developer), supportedOS(supportedOS) {}

    // Destructor
    ~Application() {}

    // Setters
    void setAppName(const std::string &appName)
    {
        this->appName = appName;
    }

    void setDeveloper(const std::string &developer)
    {
        this->developer = developer;
    }

    void setSupportedOS(const std::string &supportedOS)
    {
        this->supportedOS = supportedOS;
    }

    // Getters
    std::string getAppName() const
    {
        return appName;
    }

    std::string getDeveloper() const
    {
        return developer;
    }

    std::string getSupportedOS() const
    {
        return supportedOS;
    }

    void execute() override
    {
        printf("Application %s is running\n", appName.c_str());
    }

    void terminate() override
    {
        printf("Application %s has been terminated\n", appName.c_str());
    }

    bool upgrade() override
    {
        printf("Application %s has been upgraded\n", appName.c_str());
        return true;
    }

    bool downgrade() override
    {
        printf("Application %s has been downgraded\n", appName.c_str());
        return true;
    }
};

File: python/software/application.py
from components.software import Software
from interfaces.executable import Executable
from interfaces.upgradable import Upgradable

class Application(Software, Executable, Upgradable):
    def __init__(self, app_name="", developer="", supported_os="", version="", license_key="", size_mb=0.0):
        super().__init__(version, license_key, size_mb)
        self._app_name = app_name
        self._developer = developer
        self._supported_os = supported_os

    # Setters
    def set_app_name(self, app_name):
        self._app_name = app_name

    def set_developer(self, developer):
        self._developer = developer

    def set_supported_os(self, supported_os):
        self._supported_os = supported_os

    # Getters
    def get_app_name(self):
        return self._app_name

    def get_developer(self):
        return self._developer

    def get_supported_os(self):
        return self._supported_os

    def execute(self):
        print(f"Application {self._app_name} is running")

    def terminate(self):
        print(f"Application {self._app_name} has been terminated")

    def upgrade(self):
        print(f"Application {self._app_name} has been upgraded")
        return True

    def downgrade(self):
        print(f"Application {self._app_name} has been downgraded")
        return True